internal/controller/customresource/protection.go

```diff
@@ -79,6 +71,52 @@ func applyLastSpec(ctx context.Context, resource api.AtlasCustomResource, k8sCli
 	return nil
 }
 
+func PatchLastConfigApplied[S any](ctx context.Context, k8sClient client.Client, resource api.AtlasCustomResource, spec *S) error {
+	copy, err := setLastAppliedConfigAnnotation(resource, spec)
+	if err != nil {
+		return fmt.Errorf("failed to set the last applied config annotation: %w", err)
+	}
+
+	err = k8sClient.Patch(ctx, copy, client.MergeFrom(resource))
+	if err != nil {
+		return fmt.Errorf("failed to patch resource: %w", err)
+	}
+	return nil
+}
+
+func setLastAppliedConfigAnnotation[S any](resource api.AtlasCustomResource, spec *S) (api.AtlasCustomResource, error) {
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(spec)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get unstructured object from spec %v: %w", spec, err)
+	}
+	js, err := json.Marshal(uObj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal %v: %w", uObj, err)
+	}
+	copy := resource.DeepCopyObject().(api.AtlasCustomResource)
+	annotations := copy.GetAnnotations()
+	if annotations == nil {
```

Comment:
the nil check is not necessary, every resource in kube has annotations.

Reply:
This whole method can be as simple as:
```
func PatchLastConfigApplied[S any](ctx context.Context, k8sClient client.Client, resource api.AtlasCustomResource, spec *S) error {
	if spec == nil {
		return fmt.Errorf("spec is nil")
	}

	js, err := json.Marshal(spec)
	if err != nil {
		return fmt.Errorf("failed to marshal %v: %w", spec, err)
	}

	resourceCopy := resource.DeepCopyObject().(api.AtlasCustomResource)
	resourceCopy.GetAnnotations()[AnnotationLastAppliedConfiguration] = string(js)

	err = k8sClient.Patch(ctx, resourceCopy, client.MergeFrom(resource))
	if err != nil {
		return fmt.Errorf("failed to patch resource: %w", err)
	}
	return nil
}
```

Reply:
actually, here, a little helper method acutally makes sense if we want to preserve nil-safety:

Reply:
actually, the current method in `main` is perfectly fine, the unstructured stuff can be removed: https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/internal/controller/customresource/protection.go#L51-L80